File: app/src/main/java/com/example/aleartmycontroller/data/repository/EventRepository.kt
package com.example.aleartmycontroller.data.repository

import com.example.aleartmycontroller.BuildConfig
import com.example.aleartmycontroller.data.local.dao.EventDao
import com.example.aleartmycontroller.data.local.dao.ObservationEventDao
import com.example.aleartmycontroller.data.local.entity.EventEntity
import com.example.aleartmycontroller.data.local.entity.LOCAL_DRAFT_GOOGLE_ID_PREFIX
import com.example.aleartmycontroller.data.local.entity.isLocalDraft
import com.example.aleartmycontroller.data.remote.google.GoogleCalendarApi
import com.example.aleartmycontroller.data.remote.google.CalendarEventDateTimeRequest
import com.example.aleartmycontroller.data.remote.google.CalendarEventPatchRequest
import com.example.aleartmycontroller.data.remote.google.CalendarEventUpsertRequest
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.flow
import java.time.Instant
import java.time.OffsetDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class EventRepository @Inject constructor(
    private val eventDao: EventDao,
    private val observationEventDao: ObservationEventDao,
    private val calendarApi: GoogleCalendarApi
) {
    private val isoFormatter: DateTimeFormatter = DateTimeFormatter.ISO_OFFSET_DATE_TIME
    private val zoneId: ZoneId = ZoneId.systemDefault()

    /** 今日以降のイベントをFlowで監視（ローカルDB） */
    fun observeUpcomingEvents(): Flow<List<EventEntity>> {
        val nowMillis = System.currentTimeMillis()
        return eventDao.observeUpcoming(nowMillis)
    }

    /** WorkManager など一度だけ取得する場合に使用する */
    suspend fun getUpcomingEventsOnce(): List<EventEntity> {
        return eventDao.getUpcoming(System.currentTimeMillis())
    }

    /** 全イベントを監視（履歴画面向け） */
    fun observeAllEvents(): Flow<List<EventEntity>> = eventDao.observeAll()

    /**
     * Google Calendar APIから今日〜7日後のイベントを取得しDBへ同期する。
     * 古いキャッシュは削除する。
     */
    suspend fun syncFromCalendar() {
        val now = OffsetDateTime.now(ZoneId.systemDefault())
        val weekLater = now.plusDays(7)

        val response = calendarApi.listEvents(
            timeMin = now.format(isoFormatter),
            timeMax = weekLater.format(isoFormatter)
        )

        val entities = response.items.mapNotNull { item ->
            val startMillis = item.start?.value?.let { OffsetDateTime.parse(it, isoFormatter).toInstant().toEpochMilli() } ?: return@mapNotNull null
            val endMillis = item.end?.value?.let { OffsetDateTime.parse(it, isoFormatter).toInstant().toEpochMilli() } ?: return@mapNotNull null
            EventEntity(
                googleEventId = item.id,
                title = item.summary ?: "(無題)",
                startTime = startMillis,
                endTime = endMillis
            )
        }.associateBy { it.googleEventId }.values.toList() // 重複除去：同じ Google ID なら後勝ちで上書き

        // DB に upsert してから不要なエントリを削除
        eventDao.upsertAll(entities)
        val activeIds = entities.map { it.googleEventId }
        if (activeIds.isNotEmpty()) {
            eventDao.deleteStaleEvents(activeIds)
        }
    }

    suspend fun findById(id: Long): EventEntity? = eventDao.findById(id)

    suspend fun createDraftEvent(title: String): EventEntity {
        val now = System.currentTimeMillis()
        val draftEvent = EventEntity(
            googleEventId = "$LOCAL_DRAFT_GOOGLE_ID_PREFIX${UUID.randomUUID()}",
            title = title,
            startTime = now,
            endTime = now + 60 * 60 * 1000
        )
        val eventId = eventDao.upsert(draftEvent)
        return draftEvent.copy(eventId = eventId)
    }

    suspend fun finalizeDraftEvent(
        eventId: Long,
        endTimeMillis: Long,
        description: String?
    ): EventEntity {
        val draftEvent = eventDao.findById(eventId) ?: error("Event not found: $eventId")
        require(draftEvent.isLocalDraft()) { "Not a draft event: ${draftEvent.googleEventId}" }

        val inserted = calendarApi.insertEvent(
            body = CalendarEventUpsertRequest(
                summary = draftEvent.title,
                description = description?.takeIf { it.isNotBlank() },
                start = toRequestTime(draftEvent.startTime),
                end = toRequestTime(endTimeMillis)
            )
        )

        val finalized = draftEvent.copy(
            googleEventId = inserted.id,
            endTime = endTimeMillis
        )
        eventDao.upsert(finalized)
        observationEventDao.updateGoogleId(
            oldGoogleId = draftEvent.googleEventId,
            newGoogleId = inserted.id,
            title = finalized.title,
            startTime = finalized.startTime,
            endTime = finalized.endTime
        )
        return finalized
    }

    suspend fun appendMemoToGoogleEvent(event: EventEntity, memoText: String) {
        if (memoText.isBlank() || event.isLocalDraft()) return

        val remote = calendarApi.getEvent(eventId = event.googleEventId)
        val mergedDescription = listOfNotNull(
            remote.description?.takeIf { it.isNotBlank() },
            memoText.trim()
        ).joinToString("\n\n")

        calendarApi.patchEvent(
            eventId = event.googleEventId,
            body = CalendarEventPatchRequest(description = mergedDescription)
        )
    }

    @OptIn(kotlinx.coroutines.ExperimentalCoroutinesApi::class)
    fun observeOngoingEvent(): Flow<EventEntity?> {
        // 1分ごとに現在時刻を更新して、進行中イベントを引き直す
        val ticker = flow {
            while (true) {
                emit(System.currentTimeMillis())
                delay(60_000)
            }
        }
        return ticker.flatMapLatest { now ->
            eventDao.observeOngoing(now)
        }
    }

    suspend fun getOngoingEvent(): EventEntity? {
        return eventDao.findOngoing(System.currentTimeMillis())
    }

    private fun toRequestTime(millis: Long): CalendarEventDateTimeRequest {
        val dateTime = OffsetDateTime.ofInstant(Instant.ofEpochMilli(millis), zoneId)
            .format(isoFormatter)
        return CalendarEventDateTimeRequest(
            dateTime = dateTime,
            timeZone = zoneId.id
        )
    }
}

File: flutter/amc/lib/data/repository/event_repository.dart
// 名前付き引数 + private フィールドのため initializing formal は使えない（言語仕様）。
// ignore_for_file: prefer_initializing_formals
import 'package:drift/drift.dart' show Value;
import 'package:googleapis/calendar/v3.dart' as gcal;
import 'package:rxdart/rxdart.dart';
import 'package:synchronized/synchronized.dart';
import 'package:uuid/uuid.dart';

import '../../core/content_policy.dart';
import '../local/daos/event_dao.dart';
import '../local/daos/observation_event_dao.dart';
import '../local/database.dart';
import '../local/tables.dart';
import '../remote/google_calendar_api.dart';

/// イベントのローカル監視 + Google Calendar 同期 + 下書きライフサイクル
/// （Android: EventRepository 相当）。
class EventRepository {
  EventRepository({
    required EventDao eventDao,
    required ObservationEventDao observationEventDao,
    required GoogleCalendarApi calendarApi,
  })  : _eventDao = eventDao,
        _observationEventDao = observationEventDao,
        _calendarApi = calendarApi;

  final EventDao _eventDao;
  final ObservationEventDao _observationEventDao;
  final GoogleCalendarApi _calendarApi;

  static const _uuid = Uuid();

  /// googleEventId ごとのロック（並行メモ追記の競合防止、§9）。
  final Map<String, Lock> _memoLocks = {};
  Lock _lockFor(String googleEventId) =>
      _memoLocks.putIfAbsent(googleEventId, Lock.new);

  bool _isLocalDraft(String googleEventId) =>
      googleEventId.startsWith(localDraftGoogleIdPrefix);

  // ---- ローカル監視 ----

  Stream<List<Event>> watchUpcomingEvents() =>
      _eventDao.watchUpcoming(DateTime.now().millisecondsSinceEpoch);

  Future<List<Event>> getUpcomingEventsOnce() =>
      _eventDao.getUpcoming(DateTime.now().millisecondsSinceEpoch);

  Stream<List<Event>> watchAllEvents() => _eventDao.watchAll();

  Future<Event?> findById(int id) => _eventDao.findById(id);

  Future<Event?> getOngoingEvent() =>
      _eventDao.findOngoing(DateTime.now().millisecondsSinceEpoch);

  /// 進行中イベントを監視する。1 分ごとに現在時刻を引き直し、drift Stream に流す。
  /// （Android: 60 秒ティッカー + flatMapLatest 相当、§3.4）。
  Stream<Event?> observeOngoingEvent() {
    return Stream<int>.periodic(
      const Duration(seconds: 60),
      (_) => DateTime.now().millisecondsSinceEpoch,
    ).startWith(DateTime.now().millisecondsSinceEpoch).switchMap(
          (now) => _eventDao.watchOngoing(now),
        );
  }

  // ---- Google Calendar 同期 ----

  /// 今日〜7日後のイベントを取得して DB へ upsert し、古いキャッシュを削除する。
  Future<void> syncFromCalendar() async {
    final now = DateTime.now();
    final weekLater = now.add(const Duration(days: 7));
    final response =
        await _calendarApi.listEvents(timeMin: now, timeMax: weekLater);

    // 同じ Google ID は後勝ちで上書き（重複除去）。
    final byGoogleId = <String, EventsCompanion>{};
    for (final item in response.items ?? <gcal.Event>[]) {
      final id = item.id;
      final start = _toMillis(item.start);
      final end = _toMillis(item.end);
      if (id == null || start == null || end == null) continue;
      byGoogleId[id] = EventsCompanion.insert(
        googleEventId: id,
        title: item.summary ?? '(無題)',
        startTime: start,
        endTime: end,
      );
    }

    await _eventDao.upsertAll(byGoogleId.values.toList());
    final activeIds = byGoogleId.keys.toList();
    if (activeIds.isNotEmpty) {
      await _eventDao.deleteStaleEvents(activeIds);
    }
  }

  // ---- 突発下書き ----

  Future<Event> createDraftEvent(String title) async {
    final now = DateTime.now().millisecondsSinceEpoch;
    final googleId = '$localDraftGoogleIdPrefix${_uuid.v4()}';
    final id = await _eventDao.upsert(
      EventsCompanion.insert(
        googleEventId: googleId,
        title: title,
        startTime: now,
        endTime: now + 60 * 60 * 1000,
      ),
    );
    return (await _eventDao.findById(id))!;
  }

  Future<Event> closeDraftEvent(int eventId, int endTimeMillis) async {
    final draft = await _eventDao.findById(eventId);
    if (draft == null) throw StateError('Event not found: $eventId');
    if (!_isLocalDraft(draft.googleEventId)) {
      throw StateError('Not a draft event: ${draft.googleEventId}');
    }
    await _eventDao.upsert(_replace(draft, endTime: endTimeMillis));
    return (await _eventDao.findById(eventId))!;
  }

  /// 下書きを Google Calendar に確定し、observation_events の googleEventId を
  /// 実 ID へ付け替える(§9)。googleEventId 別ロックで保護する。
  Future<Event> finalizeDraftEvent(int eventId, String? description) async {
    final draft = await _eventDao.findById(eventId);
    if (draft == null) throw StateError('Event not found: $eventId');
    if (!_isLocalDraft(draft.googleEventId)) {
      throw StateError('Not a draft event: ${draft.googleEventId}');
    }

    return _lockFor(draft.googleEventId).synchronized(() async {
      final desc = (description != null && description.trim().isNotEmpty)
          ? description
          : null;
      final inserted = await _calendarApi.insertEvent(
        summary: draft.title,
        description: desc,
        startMillis: draft.startTime,
        endMillis: draft.endTime,
      );
      final newGoogleId = inserted.id!;
      await _eventDao.upsert(_replace(draft, googleEventId: newGoogleId));
      await _observationEventDao.updateGoogleId(
        oldGoogleId: draft.googleEventId,
        newGoogleId: newGoogleId,
        title: draft.title,
        startTime: draft.startTime,
        endTime: draft.endTime,
      );
      return (await _eventDao.findById(eventId))!;
    });
  }

  /// 既存 Google イベントの description にメモを追記（ミラー）。
  /// 下書き・空文字はスキップ。googleEventId 別ロックで保護する(§9)。
  Future<void> appendMemoToGoogleEvent(Event event, String memoText) async {
    if (memoText.trim().isEmpty || _isLocalDraft(event.googleEventId)) return;

    await _lockFor(event.googleEventId).synchronized(() async {
      final remote = await _calendarApi.getEvent(event.googleEventId);
      final parts = <String>[
        if ((remote.description ?? '').trim().isNotEmpty)
          remote.description!.trim(),
        memoText.trim(),
      ];
      final description =
          AmcContentPolicy.buildCalendarMirrorBody(parts.join('\n\n'));
      await _calendarApi.patchEvent(event.googleEventId,
          description: description);
    });
  }

  // ---- helpers ----

  /// 同じ eventId 行を全フィールド指定で置き換える（upsert は InsertMode.replace）。
  EventsCompanion _replace(
    Event base, {
    String? googleEventId,
    int? endTime,
  }) {
    return EventsCompanion(
      eventId: Value(base.eventId),
      googleEventId: Value(googleEventId ?? base.googleEventId),
      title: Value(base.title),
      startTime: Value(base.startTime),
      endTime: Value(endTime ?? base.endTime),
    );
  }

  int? _toMillis(gcal.EventDateTime? dt) {
    if (dt == null) return null;
    // 終日イベントは date のみ（dateTime が null）→ date にフォールバック(§5.1)。
    final value = dt.dateTime ?? dt.date;
    return value?.millisecondsSinceEpoch;
  }
}
